pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,448 @@
+
+	return fmt.Sprintf("%s.%s.%s", headers, payload, signature), nil
+}
+
+// Claims fills input c with JWT claims taken from payload of a token.
+func (j *jsonWebToken) Claims(c interface{}) error {
+	pBytes, err := json.Marshal(j.payload)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(pBytes, c)
+}
+
+// VerificationData build verification data which can be used for signing or verification.
+func (j *jsonWebToken) VerificationData() ([]byte, error) {
+	return jwsVerificationData(j.headers, j.payload)
+}
+
+// Header makes look up of particular header.
+func (j *jsonWebToken) Header(name string) (interface{}, bool) {
+	v, ok := j.headers[name]
+
+	return v, ok
+}
+
+// HeaderString makes look up of particular header with string value.
+func (j *jsonWebToken) HeaderString(name string) string {
+	if headerValue, ok := j.headers[name]; ok {
+		if headerStrValue, ok := headerValue.(string); ok {
+			return headerStrValue
+		}
+	}
+
+	return ""
+}
+
+// Signature returns a copy of JWT signature if defined.
+func (j *jsonWebToken) Signature() []byte {
+	if j.signature == nil {
+		return nil
+	}
+
+	return append(j.signature[:0:0], j.signature...)
+}
+
+// Signer defines JWT Signer interface. It makes signing of data and provides custom JWT headers relevant to the signer.
+type Signer interface {
+	// Sign signs.
+	Sign(data []byte) ([]byte, error)
+
+	// Headers provides JWT headers.
+	Headers() map[string]interface{}
+}
+
+// Builder construct JSON Web Token.
+type Builder interface {
+	// Signed creates JWS.
+	Signed(signer Signer) (JSONWebToken, error)
+
+	// Unsecured creates unsecured JWT.
+	Unsecured(headers map[string]interface{}) (JSONWebToken, error)
+}
+
+// builder implements Builder.
+type builder struct {
+	claims map[string]interface{}
+
+	err error
+}
+
+// Signed creates JWS.
+func (b *builder) Signed(signer Signer) (JSONWebToken, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	// build headers
+	headers := signer.Headers()
+
+	verificationData, err := jwsVerificationData(headers, b.claims)
+	if err != nil {
+		return nil, fmt.Errorf("prepare JWT verification data: %w", err)
+	}
+
+	signature, err := signer.Sign(verificationData)
+	if err != nil {
+		return nil, fmt.Errorf("sign JWT verification data: %w", err)
+	}
+
+	return &jsonWebToken{
+		headers:   headers,
+		payload:   b.claims,
+		signature: signature,
+	}, nil
+}
+
+// Unsecured creates unsecured JWT.
+func (b *builder) Unsecured(headers map[string]interface{}) (JSONWebToken, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	return &jsonWebToken{
+		headers: PrepareJWTUnsecuredHeaders(headers),
+		payload: b.claims,
+	}, nil
+}
+
+// New initiates creation of a new JSON Web Tokne. claims could be anything marshallable to the map.
+func New(claims interface{}) Builder {
+	m, err := toMap(claims)
+
+	return &builder{
+		claims: m,
+		err:    err,
+	}
+}
+
+// FromSigned parses JWS in compact serialization format.
```

Comment:
So the calling function should validate the alg header field?

Reply:
I think you meant how a caller distinguishes between `FromSigned` and `FromUnsecured`? I created these two methods because for the latter the `Verifier` is not needed. And I think that unsecured JWT is a very special case. go-jose even does not want to support that! So I think the caller does know if it's JWS or unsecured JWT.